iam_groups_authn/sync.py

```diff
@@ -18,15 +18,180 @@
+            else:
+                db = init_postgres_connection_engine(instance, updated_creds, ip_type)
+                role_service = PostgresRoleService(db)
+            logging.debug(
+                "[%s][%s] Initialized a %s connection pool."
+                % (instance, group, database_version.value)
+            )
+
+            # verify role for IAM group exists on database, create if does not exist
+            role = mysql_username(group)
+            verify_role_task = asyncio.create_task(role_service.create_group_role(role))
+
+            # get database users who have group role
+            users_with_roles_task = asyncio.create_task(
+                get_users_with_roles(role_service, role)
+            )
+
+            # await dependent tasks
+            results = await asyncio.gather(
+                add_users_task, verify_role_task, return_exceptions=True
+            )
+            # raise exception if found
+            for result in results:
+                if issubclass(type(result), Exception):
+                    raise result
+
+            # log IAM users added as database users
+            added_users = results[0]
+            logging.debug(
+                "[%s][%s] Users added to database: %s."
+                % (instance, group, list(added_users))
+            )
+
+            # revoke group role from users no longer in IAM group
+            revoke_role_task = asyncio.create_task(
+                revoke_iam_group_role(
+                    role_service,
+                    role,
+                    users_with_roles_task,
+                    group_tasks[group],
+                    database_version,
+                )
+            )
+
+            # grant group role to IAM users who are missing it on database
+            grant_role_task = asyncio.create_task(
+                grant_iam_group_role(
+                    role_service,
+                    role,
+                    users_with_roles_task,
+                    group_tasks[group],
+                    database_version,
+                )
+            )
+            results = await asyncio.gather(
+                revoke_role_task, grant_role_task, return_exceptions=True
+            )
+            # raise exception if found
+            for result in results:
+                if issubclass(type(result), Exception):
+                    raise result
+
+            # log sync info
+            revoked_users, granted_users = results
+            logging.info(
+                "[%s][%s] Sync successful: %s users were revoked group role, %s users were granted group role."
+                % (instance, group, len(revoked_users), len(granted_users))
+            )
+            logging.debug(
+                "[%s][%s] Users revoked role: %s." % (instance, group, revoked_users)
+            )
+            logging.debug(
+                "[%s][%s] Users granted role: %s." % (instance, group, granted_users)
+            )
```

Comment:
nit: it might be easier to read if this was a separate function - something like "sync group to instance"